LFS-202: Filtering tables

OrdiNeu opened the pull request:
Adds filtering support for LiveTables, visible on the user dashboard. This adds the possibility of finding e.g. forms where a patient is born before a certain year and whom are now dead.

(will squash commits before rebasing)

OrdiNeu committed:
LFS-202: Filtering tables
Add basic filtering support

OrdiNeu committed:
LFS-202: Filtering tables
Allow pagination to handle multiple filters
Refactor sanitization into its own function
Switch from storing filters as arrays of arrays to arrays of objects.

OrdiNeu committed:
LFS-202: Filtering tables
Use a deep copy from filters to active filters and back to prevent
background updates. Use default values to remove console warnings.

OrdiNeu committed:
LFS-202: Filtering tables
UI improvements

OrdiNeu committed:
LFS-202: Filtering tables
Fix comparators
Fix bug where the filter values were not being set when
the dialog was reopened.
Allow user to click on chips to edit the filters

OrdiNeu committed:
LFS-202: Filtering tables
Code style improvements

OrdiNeu committed:
LFS-202: Filtering tables
Add better support for single-questionnaire filters.

OrdiNeu committed:
LFS-202: Filtering tables
Fix issue with list selection

OrdiNeu committed:
LFS-202: Filtering tables
Fix behaviour of filters when no questionnaire is provided

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -0,0 +1,440 @@
+                +Add another filter
+              </Link>
+            </Grid>
+            <Grid item xs={3}>
```
Please use https://material-ui.com/api/dialog-actions/ to wrap the buttons.

OrdiNeu committed:
LFS-202: Filtering tables
Various UI improvements

OrdiNeu committed:
LFS-202: Filtering tables
Improve vocabulary support

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -0,0 +1,473 @@
+            color="primary"
+            onClick={saveFilters}
+            >
+            {'Save'}
```
`Apply`?

OrdiNeu committed:
LFS-202: Filtering tables
Apply comments by @marta-

anusha-dey committed:
LFS-202: Filtering tables
Display the correct filter comparators

qrst07 commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -0,0 +1,473 @@
+
+    // If this question has multiple pre-defined responses, and a displayMode='list', store the responses
+    // (we ignore list+input here because the user can select anything from those)
+    if (displayMode === 'list') {
```
I'm not sure if fields with `displayMode == 'list'` and `dataType == 'date'` (e.g. Surveilled since and Surveilled until') are handled properly - an empty list of options shows up but should a date field show up instead?

OrdiNeu replied:
Good catch! It looks like 'Surveilled Since' and 'Surveilled Until' don't actually have answer options (so I'm unsure why they have `displayMode == 'list'`). On @marta- 's advice, I'll make numeric fields always allow arbitrary input so that they can be compared using `>`, `<`, etc..

OrdiNeu committed:
LFS-202: Filtering tables
Fix the filtering of numeric fields to have precedence over the list displayMode.

OrdiNeu committed:
LFS-202: Filtering tables
Remove 'and' spacers from dialog and chips

OrdiNeu committed:
LFS-202: Filtering Tables
Fix UI spacing

OrdiNeu committed:
LFS-202: Filtering tables
Various UI improvements

OrdiNeu committed:
LFS-202: Filtering tables
Focus on the input when a field is selected.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/tableStyle.jsx:
```diff
@@ -20,6 +20,62 @@
+    selectPlaceholder: {
+        opacity: 0.3
+    },
+    verticalFieldSpacer: {
```
Is this used anywhere?

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/tableStyle.jsx:
```diff
@@ -20,6 +20,62 @@
+        width: "100%"
+    },
+    dialogPaper: {
+        maxWidth: "634px"
```
Why this magic number?

OrdiNeu committed:
LFS-202: Filtering tables
Focus user input on appropriate fields when chips are clicked or a field
is chosen.

OrdiNeu committed:
LFS-202: Filtering tables
Disable the filter dialog's portal behaviour to allow the vocab popover
to show up in the correct position

OrdiNeu committed:
LFS-202: Filtering tables
Adjust the zIndex of vocabulary questions so they behave properly with
multiple dialogs. This re-enables the portal behaviour, allowing them to
no longer force resizes of their container when they appear or
disappear.

OrdiNeu committed:
LFS-202: Filtering tables
Fix the is empty and is not empty filters.
Properly focus input on vocabulary selectors.

OrdiNeu committed:
LFS-202: Filtering tables
Commenting & cleaning code

OrdiNeu committed:
LFS-202: Filtering tables
Add automatic conversion to "is empty" if no value is provided.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -18,10 +18,13 @@
 import React, { useState } from "react";
-import { Paper, Table, TableHead, TableBody, TableRow, TableCell, TablePagination } from "@material-ui/core";
-import { Card, CardHeader, CardContent, CardActions, Typography, Button, withStyles } from "@material-ui/core";
+import { Fab, Paper, Table, TableHead, TableBody, TableRow, TableCell, TablePagination } from "@material-ui/core";
```
`Fab` doesn't appear to be used anymore as we made our own rounded (+) button.
```suggestion
import { Paper, Table, TableHead, TableBody, TableRow, TableCell, TablePagination } from "@material-ui/core";
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/tableStyle.jsx:
```diff
@@ -20,6 +20,62 @@
+        right: theme.spacing(2)
+    },
+    dialogContent: {
+    },
```
Please remove empty style.

marta- reviewed (changes requested):
Entering a `|` in any filter value will cause the request to fail with the message "fieldname and fieldvalue must have the same numberof values, as delimited by pipes (|) (500)".

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/tableStyle.jsx:
```diff
@@ -20,6 +20,62 @@
+    },
+    categoryField: {
+        width: "100%",
+    },
```
`answerField`, `categoryField` and `comparatorSelector` all have the same styling: `width: 100%'. I suggest using a CSS class for all that applies this style.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -79,6 +83,19 @@ function LiveTable(props) {
+    let filters = newPage.filters || cachedFilters;
+    if (filters != null) {
+      url.searchParams.set("joinchildren", joinChildren);
+      url.searchParams.set("filternames", filters["fields"].join("|"));
```
Instead of joining all parameters into a string, try passing them individually under the `filternames` query param:

`filters.fields.forEach(i => url.searchParams.append('filternames', i));` (via @sdumitriu ).

Same for comparators, values, etc.

The servlet will receive them each parameter as an array.

OrdiNeu committed:
LFS-202: Filtering tables
Improve support for multiple filters by not using arbitrary seperators.

OrdiNeu committed:
LFS-202: Filtering tables
Properly grab the focus for list entries

OrdiNeu committed:
LFS-202: Filtering tables
Apply suggestions by @marta-

OrdiNeu committed:
LFS-202: Filtering tables
Remove unused classes.

marta- reviewed (approved):
Looks fine to me for now. I created [LFS-244](https://phenotips.atlassian.net/browse/LFS-244) and [LFS-245](https://phenotips.atlassian.net/browse/LFS-245) for future improvements.
